plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -0,0 +1,860 @@
+}
+
+static FuFirmware *
+fu_bcm57xx_device_read_firmware (FuDevice *device, GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_bcm57xx_firmware_new ();
+	g_autoptr(GBytes) fw = NULL;
+
+	/* read from hardware */
+	fw = fu_bcm57xx_device_dump_firmware (device, error);
+	if (fw == NULL)
+		return NULL;
+	if (!fu_firmware_parse (firmware, fw, FWUPD_INSTALL_FLAG_NONE, error))
+		return NULL;
+
+	/* remove images that will contain user-data */
+	if (!fu_firmware_remove_image_by_id (firmware, "info", error))
+		return NULL;
+	if (!fu_firmware_remove_image_by_id (firmware, "info2", error))
+		return NULL;
+	if (!fu_firmware_remove_image_by_id (firmware, "vpd", error))
+		return NULL;
+	return g_steal_pointer (&firmware);
+}
+
+static FuFirmware *
+fu_bcm57xx_device_prepare_firmware (FuDevice *device,
+				    GBytes *fw,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	g_autoptr(FuFirmware) firmware = NULL;
+	g_autoptr(FuFirmware) firmware_tmp = fu_bcm57xx_firmware_new ();
+	g_autoptr(FuFirmwareImage) img_ape = NULL;
+	g_autoptr(FuFirmwareImage) img_stage1 = NULL;
+	g_autoptr(FuFirmwareImage) img_stage2 = NULL;
+
+	/* try to parse NVRAM, stage1 or APE */
+	if (!fu_firmware_parse (firmware_tmp, fw, flags, error))
+		return NULL;
+
+	/* for full NVRAM image, verify if correct device */
+	if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
+		guint16 vid = fu_bcm57xx_firmware_get_vendor (FU_BCM57XX_FIRMWARE (firmware_tmp));
+		guint16 did = fu_bcm57xx_firmware_get_model (FU_BCM57XX_FIRMWARE (firmware_tmp));
+		if (vid != 0x0 && did != 0x0 &&
+		    (fu_udev_device_get_vendor (FU_UDEV_DEVICE (device)) != vid ||
+		     fu_udev_device_get_model (FU_UDEV_DEVICE (device)) != did)) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "PCI vendor or model incorrect, got: %04X:%04X",
+				     vid, did);
+			return NULL;
+		}
+	}
+
+	/* get the existing firmware from the device */
+	firmware = fu_bcm57xx_device_read_firmware (device, error);
+	if (firmware == NULL)
+		return NULL;
+
+	/* merge in all the provided images into the existing firmware */
+	img_stage1 = fu_firmware_get_image_by_id (firmware_tmp, "stage1", NULL);
+	if (img_stage1 != NULL)
+		fu_firmware_add_image (firmware, img_stage1);
+	img_stage2 = fu_firmware_get_image_by_id (firmware_tmp, "stage2", NULL);
+	if (img_stage2 != NULL)
+		fu_firmware_add_image (firmware, img_stage2);
+	img_ape = fu_firmware_get_image_by_id (firmware_tmp, "ape", NULL);
+	if (img_ape != NULL)
+		fu_firmware_add_image (firmware, img_ape);
+
+	/* success */
+	return g_steal_pointer (&firmware);
+}
+
+static gboolean
+fu_bcm57xx_device_write_firmware (FuDevice *device,
+				  FuFirmware *firmware,
+				  FwupdInstallFlags flags,
+				  GError **error)
+{
+	FuBcm57xxDevice *self= FU_BCM57XX_DEVICE (device);
+	const guint8 *buf;
+	gsize bufsz = 0;
+	gsize bufsz_dwrds = fu_device_get_firmware_size_max (FU_DEVICE (self)) / sizeof(guint32);
```

Comment:
If fu_device_get_firmware_size_max is ever something larger than the actual nvm/eeprom size in hardware, this could cause a wrap-around in hardware and cause garbage to be written to eeprom. Maybe it'd be better to only write the size of the firmware image provided in the update.cab file? This would mean that a small firmware update image would result in a portion of the eeprom being unmodified.

(note: I've update the bcm5719-fw repo to  generate an image that's as small as possible).

Reply:
> only write the size of the firmware image provided in the update.cab file

Indeed, good catch. Fixed.